src/kernel.cr

```diff
@@ -125,22 +125,47 @@ end
 module AtExitHandlers
   @@running = false
 
+  private class_getter(handlers) { [] of Int32 -> }
+  private class_getter(exceptions) { [] of Exception }
+
   def self.add(handler)
-    handlers = @@handlers ||= [] of Int32 ->
+    raise "Cannot use at_exit from an at_exit handler" if @@running
+
     handlers << handler
   end
 
+  def self.add_exception(ex : Exception)
+    # at_exit handlers can't be cleared now, but might in the future.
+    # The exceptions are stored in a separate list to make sure it won't be cleared.
+
+    exceptions << ex
+  end
+
   def self.run(status)
-    return if @@running
     @@running = true
 
-    @@handlers.try &.reverse_each do |handler|
-      begin
-        handler.call status
-      rescue handler_ex
-        STDERR.puts "Error running at_exit handler: #{handler_ex}"
+    if handlers = @@handlers
+      # Run the registered handlers in reverse order
+      while handler = handlers.pop?
+        begin
```

Comment:
IIRC you can skip `begin ... end` block altogether:

```cr
while handler = handlers.pop?
  # ...
rescue handler_ex
  # ...
end
```

Reply:
I thought I worked too, but it doesn't inside a while..

Reply:
I don't remember exactly but I think it was ambiguous whether the rescue should be after each loop, or after the whole while.

Reply:
yep